rasa/graph_components/validators/finetuning_validator.py

```diff
@@ -0,0 +1,341 @@
+    Note that even though conditions 2. and 3. differ based on which part we finetune,
+    condition 1. always covers both parts, i.e. NLU and Core.
+    """
+
+    FILENAME = "fingerprints-for-validation.json"
+
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+        fingerprints: Optional[Dict[Text, Text]] = None,
+    ) -> None:
+        """Instantiates a `FineTuningValidator`.
+
+        Args:
+            model_storage: Storage which graph components can use to persist and load
+                themselves.
+            resource: Resource locator for this component which can be used to persist
+                and load itself from the `model_storage`.
+            execution_context: Information about the current graph run.
+            fingerprints: a dictionary of fingerprints generated by a
+                `FineTuningValidator`
+        """
+        self._execution_context = execution_context
+        self._model_storage = model_storage
+        self._resource = resource
+        self._fingerprints: Dict[Text, Text] = fingerprints or {}
+
+    def validate_nlu_only(
+        self, importer: TrainingDataImporter, config: Dict[Text, Any]
+    ) -> None:
+        """Validates whether we can finetune the NLU part when finetuning is enabled.
+
+        Args:
+            importer: a training data importer
+            config: the configuration (including the recipe)
+
+        Raises:
+            `InvalidConfigException` if there is a conflict
+        """
+        self._validate(importer=importer, config=config, nlu=True, core=False)
+
+    def validate_core_only(
+        self, importer: TrainingDataImporter, config: Dict[Text, Any]
+    ) -> None:
+        """Validates whether we can finetune the Core part when finetuning is enabled.
+
+        Args:
+            importer: a training data importer
+            config: the configuration (including the recipe)
+
+        Raises:
+            `InvalidConfigException` if there is a conflict
+        """
+        self._validate(importer=importer, config=config, nlu=False, core=True)
+
+    def validate(
+        self, importer: TrainingDataImporter, config: Dict[Text, Any],
+    ) -> None:
+        """Validates whether we can finetune Core and NLU when finetuning is enabled.
+
+        Args:
+            importer: a training data importer
+            config: the configuration (including the recipe)
+
+        Raises:
+            `InvalidConfigException` if there is a conflict
+        """
+        self._validate(importer, config, nlu=True, core=True)
+
+    def _validate(
+        self,
+        importer: TrainingDataImporter,
+        config: Dict[Text, Any],
+        nlu: bool = True,
+        core: bool = True,
+    ) -> None:
+        """Validate whether the finetuning setting conflicts with other settings.
+
+        Note that this validation always takes into account the configuration of
```

Comment:
I feel this is implementation details and does not need to be in a docstring.

Reply:
I found this a bit confusing which is why I added it explicitly. If I want to finetune the NLU model and than re-train the core part with that fine tuned model, then I must adapt the core part only after finetuning (even though that should have no effect on the nlu part), no?

Reply:
ohhhh I see what you mean now. Then yes I agree, this makes sense to keep.